--- asterisk/channels/h323/Makefile.in.orig	2012-06-21 15:23:35.000000000 +0600
+++ asterisk/channels/h323/Makefile.in	2012-06-21 17:02:28.000000000 +0600
@@ -4,45 +4,27 @@
 # Make file for OpenH323 support layer
 #
 
-.PHONY: Makefile.ast clean
+.PHONY:		Makefile.ast
 
-default::	@OPENH323_BUILD@
-
-# Verify those options with main Makefile
-STDCCFLAGS	= -DNDEBUG
+STDCCFLAGS      := -DPTRACING
 STDCCFLAGS	+= -I../../include -include ../../include/asterisk/autoconfig.h
 STDCCFLAGS	+= -fPIC
-#OPTCCFLAGS	+=
-CFLAGS		= -pipe
-TARGET		= libchanh323.a
-TARGET		+= Makefile.ast
-SOURCES		= ast_h323.cxx compat_h323.cxx cisco-h225.cxx caps_h323.cxx
-OBJDIR		= .
-OBJS		=
+CFLAGS		= -pipe -fexceptions
+SOURCES		:= ast_h323.cxx compat_h323.cxx cisco-h225.cxx caps_h323.cxx
+OBJDIR		:= obj_linux_x86@OPENH323_SUFFIX@
+OBJS		:= ast_h323.o compat_h323.o cisco-h225.o caps_h323.o
+BUILD_TYPE	:= @OPENH323_BUILD@
+ifeq ($(BUILD_TYPE),debug)
+DEBUG		:= 1
+endif
 
 ifndef OPENH323DIR
 OPENH323DIR=@OPENH323DIR@
 endif
 
-ifneq ($(wildcard $(OPENH323DIR)/openh323u.mak),)
-include $(OPENH323DIR)/openh323u.mak
-endif
-
-notrace::
-	$(MAKE) NOTRACE=1 opt
-
-$(SOURCES)::	Makefile ../../Makefile
-	touch $@
-
-libchanh323.a:	$(OBJS)
-	ar crv $@ $(OBJS)
-
-# 
-# We have this file in svn, so this is commented out to ensure it doesn't try
-# to run implicitly.  However, it's still here for reference.
-#
-#cisco-h225.cxx::	cisco-h225.asn
-#	asnparser -m CISCO_H225 -c $<
+libchanh323.a:	Makefile.ast $(addprefix $(OBJDIR)/,$(OBJS))
+	make @OPENH323_BUILD@
+	ar crv $@ $(OBJDIR)/*.o
 
 Makefile.ast:
 	@echo H323CFLAGS  = $(STDCCFLAGS) $(OPTCCFLAGS) $(CFLAGS) >$@.tmp
@@ -51,3 +33,10 @@
 	@if [ -r $@ ] && cmp -s $@ $@.tmp; then rm -f $@.tmp; else mv -f $@.tmp $@; fi
 
 clean::
+	rm -vf libchanh323.a
+	rm -vf Makefile.ast
+
+dist-clean::
+	rm -vf Makefile
+
+include $(OPENH323DIR)/openh323u.mak
