--- asterisk/include/asterisk/autoconfig.h.in.orig	2012-10-19 02:02:02.000000000 +0600
+++ asterisk/include/asterisk/autoconfig.h.in	2013-03-14 15:55:59.000000000 +0600
@@ -1220,6 +1220,11 @@
 /* Define to 1 if running on Darwin. */
 #undef _DARWIN_UNLIMITED_SELECT
 
+/* Enable large inode numbers on Mac OS X 10.5.  */
+#ifndef _DARWIN_USE_64_BIT_INODE
+# define _DARWIN_USE_64_BIT_INODE 1
+#endif
+
 /* Number of bits in a file offset, on hosts where this is settable. */
 #undef _FILE_OFFSET_BITS
 
